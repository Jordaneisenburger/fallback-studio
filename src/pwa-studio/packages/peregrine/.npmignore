.eslintrc.js
/lib/**/__{docs,helpers,mocks,tests}__/**
/scripts
