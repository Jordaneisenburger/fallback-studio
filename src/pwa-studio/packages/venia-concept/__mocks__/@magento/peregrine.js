import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useApolloContext = jest.fn(peregrine.useApolloContext);
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useQuery = jest.fn(peregrine.useQuery);
const useQueryResult = jest.fn(peregrine.useQueryResult);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useApolloContext,
    useEventListener,
    useDropdown,
    useQuery,
    useQueryResult,
    useSearchParam,
    useWindowSize,

    useToasts
};
