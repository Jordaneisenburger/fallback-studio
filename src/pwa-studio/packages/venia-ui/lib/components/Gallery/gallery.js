import React, { useMemo } from 'react';
import { string, shape, array } from 'prop-types';

import { mergeClasses } from '../../classify';
import GalleryItem from './item';
import defaultClasses from './gallery.css';

// map Magento 2.3.1 schema changes to Venia 2.0.0 proptype shape to maintain backwards compatibility
const mapGalleryItem = item => {
    const { small_image } = item;
    return {
        ...item,
        small_image:
            typeof small_image === 'object' ? small_image.url : small_image
    };
};

/**
 * Renders a Gallery of items. If items is an array of nulls Gallery will render
 * a placeholder item for each.
 *
 * @params {Array} props.items an array of items to render
 */
const Gallery = props => {
    const classes = mergeClasses(defaultClasses, props.classes);

    const { items } = props;

    const galleryItems = useMemo(
        () =>
            items.map((item, index) => {
                if (item === null) {
                    return <GalleryItem key={index} />;
                }
                return <GalleryItem key={index} item={mapGalleryItem(item)} />;
            }),
        [items]
    );

    return (
        <div className={classes.root}>
            <div className={classes.items}>{galleryItems}</div>
        </div>
    );
};

Gallery.propTypes = {
    classes: shape({
        filters: string,
        items: string,
        pagination: string,
        root: string
    }),
    items: array.isRequired
};

export default Gallery;
