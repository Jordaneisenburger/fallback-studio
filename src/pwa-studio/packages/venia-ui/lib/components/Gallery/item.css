.root {
}

.images {
    display: grid;
    grid-template-areas: 'main';
    margin-bottom: 0.65rem;
}

.imageContainer {
    grid-area: main;
}

.image {
    display: block;
    height: auto;
    opacity: 1;
    transition-duration: 512ms;
    transition-property: opacity, visibility;
    transition-timing-function: ease-out;
    visibility: visible;
    width: 100%;
}

.imagePlaceholder {
    composes: image;
    background-color: rgb(var(--venia-grey));
}

.name,
.price {
    font-size: 0.875rem;
    line-height: 1rem;
    margin: 0.5rem 0;
    min-height: 1rem;
}

/* state: pending */

.root_pending {
    composes: root;
}

.image_pending {
    composes: image;
    background-color: rgb(var(--venia-grey));
}

.images_pending {
    composes: images;
}

.name_pending {
    composes: name;
    background-color: rgb(var(--venia-grey));
}

.price_pending {
    composes: price;
    background-color: rgb(var(--venia-grey));
    width: 3rem;
}
