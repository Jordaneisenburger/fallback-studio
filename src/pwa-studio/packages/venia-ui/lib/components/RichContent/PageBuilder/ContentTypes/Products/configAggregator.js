import { getAdvanced } from '../../utils';

export default (node, props) => {
    const forms = node.querySelectorAll('[data-product-sku]');
    let carouselConfig = {};

    if (props.appearance === 'carousel') {
        carouselConfig = {
            autoplay: node.getAttribute('data-autoplay') === 'true',
            autoplaySpeed: parseInt(node.getAttribute('data-autoplay-speed')),
            infinite: node.getAttribute('data-infinite-loop') === 'true',
            arrows: node.getAttribute('data-show-arrows') === 'true',
            dots: node.getAttribute('data-show-dots') === 'true',
            carouselMode: node.getAttribute('data-carousel-mode'),
            centerPadding: node.getAttribute('data-center-padding')
        };
    }

    return {
        skus: [...forms].map(form => form.getAttribute('data-product-sku')),
        ...carouselConfig,
        ...getAdvanced(node)
    };
};
