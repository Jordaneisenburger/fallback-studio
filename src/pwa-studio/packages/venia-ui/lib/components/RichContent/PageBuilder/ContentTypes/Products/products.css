.root {
}
.root a {
    text-decoration: none;
}

.error {
    text-align: center;
    padding: 2rem 0;
}

.galleryItems {
    composes: items from '../../../../Gallery/gallery.css';
    grid-template-columns: repeat(5, 1fr);
}

@media (max-width: 640px) {
    .galleryItems {
        grid-template-columns: repeat(2, 1fr);
    }
}
