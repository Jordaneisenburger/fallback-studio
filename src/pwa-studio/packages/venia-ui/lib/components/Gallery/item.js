import React from 'react';
import { string, number, shape } from 'prop-types';
import { Link, resourceUrl } from '@magento/venia-drivers';
import { Price } from '@magento/peregrine';
import { transparentPlaceholder } from '@magento/peregrine/lib/util/images';
import { UNCONSTRAINED_SIZE_KEY } from '@magento/peregrine/lib/talons/Image/useImage';

import { mergeClasses } from '../../classify';
import Image from '../Image';
import defaultClasses from './item.css';

// The placeholder image is 4:5, so we should make sure to size our product
// images appropriately.
const IMAGE_WIDTH = 300;
const IMAGE_HEIGHT = 375;

// Gallery switches from two columns to three at 640px.
const IMAGE_WIDTHS = new Map()
    .set(640, IMAGE_WIDTH)
    .set(UNCONSTRAINED_SIZE_KEY, 840);

// TODO: get productUrlSuffix from graphql when it is ready
const productUrlSuffix = '.html';

const ItemPlaceholder = ({ classes }) => (
    <div className={classes.root_pending}>
        <div className={classes.images_pending}>
            <Image
                alt="Placeholder for gallery item image"
                classes={{
                    image: classes.image_pending,
                    root: classes.imageContainer
                }}
                src={transparentPlaceholder}
            />
        </div>
        <div className={classes.name_pending} />
        <div className={classes.price_pending} />
    </div>
);

const GalleryItem = props => {
    const { item } = props;

    const classes = mergeClasses(defaultClasses, props.classes);

    if (!item) {
        return <ItemPlaceholder classes={classes} />;
    }

    const { name, price, small_image, url_key } = item;
    const productLink = resourceUrl(`/${url_key}${productUrlSuffix}`);

    return (
        <div className={classes.root}>
            <Link to={productLink} className={classes.images}>
                <Image
                    alt={name}
                    classes={{
                        image: classes.image,
                        root: classes.imageContainer
                    }}
                    height={IMAGE_HEIGHT}
                    resource={small_image}
                    widths={IMAGE_WIDTHS}
                />
            </Link>
            <Link to={productLink} className={classes.name}>
                <span>{name}</span>
            </Link>
            <div className={classes.price}>
                <Price
                    value={price.regularPrice.amount.value}
                    currencyCode={price.regularPrice.amount.currency}
                />
            </div>
        </div>
    );
};

GalleryItem.propTypes = {
    classes: shape({
        image: string,
        imageContainer: string,
        imagePlaceholder: string,
        image_pending: string,
        images: string,
        images_pending: string,
        name: string,
        name_pending: string,
        price: string,
        price_pending: string,
        root: string,
        root_pending: string
    }),
    item: shape({
        id: number.isRequired,
        name: string.isRequired,
        small_image: string.isRequired,
        url_key: string.isRequired,
        price: shape({
            regularPrice: shape({
                amount: shape({
                    value: number.isRequired,
                    currency: string.isRequired
                }).isRequired
            }).isRequired
        }).isRequired
    })
};

export default GalleryItem;
