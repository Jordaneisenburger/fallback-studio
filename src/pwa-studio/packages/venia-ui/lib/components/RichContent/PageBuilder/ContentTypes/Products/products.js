import React from 'react';
import defaultClasses from './products.css';
import { mergeClasses } from '../../../../../classify';
import { arrayOf, shape, string } from 'prop-types';
import Gallery from '../../../../Gallery';
import GET_PRODUCTS_BY_SKU from '../../../../../queries/getProductsBySku.graphql';
import { useQuery } from '@apollo/react-hooks';

/**
 * Sort products based on the original order of SKUs
 *
 * @param {Array} skus
 * @param {Array} products
 * @returns {Array}
 */
const restoreSortOrder = (skus, products) => {
    const productsBySku = new Map();
    products.forEach(product => {
        productsBySku.set(product.sku, product);
    });
    return skus.map(sku => productsBySku.get(sku));
};

/**
 * Page Builder Products component.
 *
 * This component is part of the Page Builder / PWA integration. It can be consumed without Page Builder.
 *
 * @typedef Products
 * @kind functional component
 *
 * @param {props} props React component props
 *
 * @returns {React.Element} A React component that displays a Products based on a number of skus.
 */
const Products = props => {
    const classes = mergeClasses(defaultClasses, props.classes);
    const {
        skus,
        textAlign,
        border,
        borderColor,
        borderWidth,
        borderRadius,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        paddingTop,
        paddingRight,
        paddingBottom,
        paddingLeft,
        cssClasses = []
    } = props;

    const dynamicStyles = {
        textAlign,
        border,
        borderColor,
        borderWidth,
        borderRadius,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        paddingTop,
        paddingRight,
        paddingBottom,
        paddingLeft
    };

    const { loading, error, data } = useQuery(GET_PRODUCTS_BY_SKU, {
        variables: { skus }
    });

    if (loading) return null;

    if (error || data.products.items.length === 0) {
        return (
            <div
                style={dynamicStyles}
                className={[classes.root, ...cssClasses].join(' ')}
            >
                <div className={classes.error}>{'No products to display'}</div>
            </div>
        );
    }

    return (
        <div
            style={dynamicStyles}
            className={[classes.root, ...cssClasses].join(' ')}
        >
            <Gallery
                items={restoreSortOrder(skus, data.products.items)}
                classes={{ items: classes.galleryItems }}
            />
        </div>
    );
};

/**
 * Props for {@link Products}
 *
 * @typedef props
 *
 * @property {Object} classes An object containing the class names for the Products
 * @property {String} classes.root CSS class for products
 * @property {String} classes.galleryItems CSS class to modify child gallery items
 * @property {String} classes.error CSS class for displaying fetch errors
 * @property {Array} skus List of SKUs to load into product list
 * @property {String} textAlign Alignment of content within the products list
 * @property {String} border CSS border property
 * @property {String} borderColor CSS border color property
 * @property {String} borderWidth CSS border width property
 * @property {String} borderRadius CSS border radius property
 * @property {String} marginTop CSS margin top property
 * @property {String} marginRight CSS margin right property
 * @property {String} marginBottom CSS margin bottom property
 * @property {String} marginLeft CSS margin left property
 * @property {String} paddingTop CSS padding top property
 * @property {String} paddingRight CSS padding right property
 * @property {String} paddingBottom CSS padding bottom property
 * @property {String} paddingLeft CSS padding left property
 * @property {Array} cssClasses List of CSS classes to be applied to the component
 */
Products.propTypes = {
    classes: shape({
        root: string,
        galleryItems: string,
        error: string
    }),
    skus: arrayOf(string),
    textAlign: string,
    border: string,
    borderColor: string,
    borderWidth: string,
    borderRadius: string,
    marginTop: string,
    marginRight: string,
    marginBottom: string,
    marginLeft: string,
    paddingTop: string,
    paddingRight: string,
    paddingBottom: string,
    paddingLeft: string,
    cssClasses: arrayOf(string)
};

export default Products;
