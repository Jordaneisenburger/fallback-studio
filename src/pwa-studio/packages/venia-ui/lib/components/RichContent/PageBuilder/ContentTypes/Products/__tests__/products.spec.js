import { createTestInstance } from '@magento/peregrine';
import React from 'react';
import Products from '../products';

jest.mock('@apollo/react-hooks', () => ({
    useQuery: jest.fn()
}));
import { useQuery } from '@apollo/react-hooks';
jest.mock('../../../../../Gallery', () => jest.fn());
import Gallery from '../../../../../Gallery';
const mockGallery = Gallery.mockImplementation(() => 'Gallery');

test('render products with no props & no products', () => {
    useQuery.mockImplementation(() => {
        return {
            data: {
                products: {
                    items: []
                }
            },
            error: false,
            loading: false
        };
    });

    const component = createTestInstance(<Products />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('render products with all props & no products', () => {
    useQuery.mockImplementation(() => {
        return {
            data: {
                products: {
                    items: []
                }
            },
            error: false,
            loading: false
        };
    });

    const productsProps = {
        skus: ['TEST-1', 'TEST-2'],
        textAlign: 'right',
        border: 'solid',
        borderColor: 'red',
        borderWidth: '10px',
        borderRadius: '15px',
        marginTop: '10px',
        marginRight: '10px',
        marginBottom: '10px',
        marginLeft: '10px',
        paddingTop: '10px',
        paddingRight: '10px',
        paddingBottom: '10px',
        paddingLeft: '10px',
        cssClasses: ['test-class']
    };

    const component = createTestInstance(<Products {...productsProps} />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('render products with loading state', () => {
    useQuery.mockImplementation(() => {
        return {
            data: {
                products: {
                    items: []
                }
            },
            error: false,
            loading: true
        };
    });

    const component = createTestInstance(<Products />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('render products with error state', () => {
    useQuery.mockImplementation(() => {
        return {
            data: {
                products: {
                    items: []
                }
            },
            error: true,
            loading: false
        };
    });

    const component = createTestInstance(<Products />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('render products and ensure order is correct passed to Gallery', () => {
    useQuery.mockImplementation(() => {
        return {
            data: {
                products: {
                    items: [
                        {
                            sku: 'TEST-2'
                        },
                        {
                            sku: 'TEST-1'
                        }
                    ]
                }
            },
            error: false,
            loading: false
        };
    });

    const productProps = {
        skus: ['TEST-1', 'TEST-2']
    };

    createTestInstance(<Products {...productProps} />);
    expect(mockGallery).toHaveBeenCalledWith(
        expect.objectContaining({
            items: [
                {
                    sku: 'TEST-1'
                },
                {
                    sku: 'TEST-2'
                }
            ]
        }),
        expect.anything()
    );
});
