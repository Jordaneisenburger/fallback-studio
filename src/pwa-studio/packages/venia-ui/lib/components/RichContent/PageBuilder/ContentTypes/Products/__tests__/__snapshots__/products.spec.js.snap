// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render products with all props & no products 1`] = `
<div
  className=" test-class"
  style={
    Object {
      "border": "solid",
      "borderColor": "red",
      "borderRadius": "15px",
      "borderWidth": "10px",
      "marginBottom": "10px",
      "marginLeft": "10px",
      "marginRight": "10px",
      "marginTop": "10px",
      "paddingBottom": "10px",
      "paddingLeft": "10px",
      "paddingRight": "10px",
      "paddingTop": "10px",
      "textAlign": "right",
    }
  }
>
  <div>
    No products to display
  </div>
</div>
`;

exports[`render products with error state 1`] = `
<div
  className=""
  style={
    Object {
      "border": undefined,
      "borderColor": undefined,
      "borderRadius": undefined,
      "borderWidth": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "textAlign": undefined,
    }
  }
>
  <div>
    No products to display
  </div>
</div>
`;

exports[`render products with loading state 1`] = `null`;

exports[`render products with no props & no products 1`] = `
<div
  className=""
  style={
    Object {
      "border": undefined,
      "borderColor": undefined,
      "borderRadius": undefined,
      "borderWidth": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "textAlign": undefined,
    }
  }
>
  <div>
    No products to display
  </div>
</div>
`;
